core/src/main/java/jenkins/security/UserDetailsCache.java

```diff
@@ -177,12 +177,13 @@ public UserDetails call() throws Exception {
                 }
                 existenceCache.put(this.idOrFullName, Boolean.TRUE);
                 return userDetails;
+            } catch (UserMayOrMayNotExistException | DataAccessException e) {
```

Comment:
When `UserMayOrMayNotExistException` was thrown, we were flagging the user as non existent in the cache. And the next time impersonation was requested for the user, it would result in a different exception [UsernameNotFoundException](https://github.com/jenkinsci/jenkins/blob/stable-2.222/core/src/main/java/jenkins/security/UserDetailsCache.java#L124) that is handled differently by the impersonation [User#getUserDetailsForImpersonation](https://github.com/jenkinsci/jenkins/blob/stable-2.222/core/src/main/java/hudson/model/User.java#L392). 

It seems inconsistent and was the reason for the tests to fail (when using the cache the `UsernameNotFoundException` would make the impersonation fail whereas when no cache is used the `UserMayOrMayNotExistException` is raised and the impersonation does not fail).

I feel like this is a correct fix, it may however impact the efficiency of the cache for user lookup as it may discarding request for both `UsernameNotFoundException` and the subclass `UserMayOrMayNotExistException`. So with this change, the existence cache hit ratio will be reduced. I need a review on this.

Note: The method [UserDetailsCache#getCached](https://github.com/jenkinsci/jenkins/blob/stable-2.222/core/src/main/java/jenkins/security/UserDetailsCache.java#L102) throw the exception `UserMayOrMayNotExistException`. Maybe this is a method that we could use for this improvement.

Reply:
This change makes sense. It results in the cache containing a more accurate representation of the known state. If the UserMayOrMayNotExist it doesn't make sense to populate the cache with a statement that the user doesn't exist.

It's true that this may change the cache hit ratio. For at least some of the more popular providers with a recommended configuration this is probably an unusual configuration. With providers that routinely throw UserMayOrMayNotExistException this could result in observably different cache behavior. For example, with LDAP without the configuration to search for users.

Still, this results in a more accurate cache, without erroneously declaring that the user does not exist. It's possible that the less accurate cache is sufficient for the existing use, while insufficient for the new one. Changing it could make the existing use slower while not adding data meaningful for that case.

No matter how I look at it I conclude that this proposal is better by more correctly representing the state.